### ISSUE TITLE
False positive from failing to narrow union type across testing of same condition

### ISSUE THREAD

**Issue by jordan-schneider** [python/mypy#11016]
**Bug Report**

There is a valid type narrowing inference that is not being performed when the same conditional is checked in two places, the type is narrowed in the first condition, and then the variable is used in the second. See example.

**To Reproduce**

```python
from typing import Optional
switch = True
x: Optional[int] = 5

if switch:
  assert x is not None

# Other computations that don't reassign x

if switch:
  # "Unsupported operand types for + ("None" and "int")
  # Left operand is of type "Optional[int]"
  y = x + 5

```

**Expected Behavior**

No warning.

**Actual Behavior**

A false positive type error

**Your Environment**

- Mypy version used: mypy 0.910
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: Python 3.8.2
- Operating system and version: Ubuntu bionic 18.04

This comes up a lot when I have two arguments in a function, one that decides if a feature is enabled, and another that is a value that is only needed when that feature is enabled. I want to check that both values are provided on entry, but only use the optional argument at the end e,g,

```python
def f (..., write_output: bool, outdir: Optional[str] = None):
  if write_output:
    assert outdir is not None
  ...
  if write_output:
    write(obj, outdir)
```

**Comment by hauntsaninja** [comment]
Unfortunately, mypy isn't built to understand the relationship between variable types and this would be very hard to implement (I don't think any other type checkers can do this either).

In some cases, using a conditional context manager can be a nice way to do this stuff (e.g. `ctx = ... if condition else contextlib.nullcontext`). You can always double up your asserts as well.
